### ISSUE TITLE
Poor performance for Conv2D using "same" padding scheme

### ISSUE THREAD

**Issue by kelvinlyy** [keras-team/keras#15171]
**System information**.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.8

**Describe the current behavior**.
When we set the "padding" scheme of tf.keras.layers.Conv2D as "same", the computation will usually take abnormally longer time (usually more than double) than the "valid" "padding" scheme with the same parameters.

![image](https://user-images.githubusercontent.com/51120508/129440977-c74ad889-6eaf-4345-b757-bfa1f2349ab6.png)

The issue will be more severe when we set filter number to be a larger time.

**Describe the expected behavior**.
Although "same" "padding" scheme is expected to cost more resources to run, I am not confident whether the large time difference is acceptable

**Standalone code to reproduce the issue**.

https://colab.research.google.com/drive/19rK3voM4605PVyl1Sh91Xd-Ojyl1_1qK?usp=sharing#scrollTo=hHDoEjOF16mG

**Comment by sushreebarsa** [comment]
@kelvinlyy sorry for the late response! 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by kelvinlyy** [comment]
> @kelvinlyy sorry for the late response!
> In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

Hi, the code snippet for reproducing the issue is as below:

```
# initialize input_x
input_x = tf.random.normal((50000, 32, 32, 3))
```

```
# cov2d function with 'valid' padding
tf_conv2d_layer_valid = tf.keras.layers.Conv2D(filters=1, kernel_size=10, strides=(1,1), padding='valid', dilation_rate = (1,1))

# measure time taken for 'valid' padding conv2d 
start_time = time.time()
result = tf_conv2d_layer_valid(input_x)
end_time = time.time()

print("Time taken:", end_time-start_time)
```

```
# cov2d function with 'same' padding
tf_conv2d_layer_same = tf.keras.layers.Conv2D(filters=1, kernel_size=10, strides=(1,1), padding='same', dilation_rate = (1,1))

# measure time taken for 'same' padding conv2d 
start_time = time.time()
result = tf_conv2d_layer_same(input_x)
end_time = time.time()

print("Time taken:", end_time-start_time)
```

**Comment by sushreebarsa** [comment]
@kelvinlyy Thanks for the update! Could you please check with the latest stable version of TF 2.6.0 ,less time difference is observed while running your code on colab ?Please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/cb415d8ae3f9d968ad60acb6e1f13352/conv2d_padding.ipynb)for reference & let us know if it helps?Thank you!

**Comment by sushreebarsa** [comment]
@kelvinlyy Thanks for the update! Could you please check with the latest stable version of TF 2.6.0 ,less time difference is observed while running your code on colab ?Please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/cb415d8ae3f9d968ad60acb6e1f13352/conv2d_padding.ipynb)for reference & let us know if it helps?Thank you!

**Comment by sushreebarsa** [comment]
@kelvinlyy Thanks for the update! Could you please check with the latest stable version of TF 2.6.0 ,less time difference is observed while running your code on colab ?Please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/cb415d8ae3f9d968ad60acb6e1f13352/conv2d_padding.ipynb)for reference & let us know if it helps?Thank you!

**Comment by kelvinlyy** [comment]
> 
> 
> @kelvinlyy Thanks for the update! Could you please check with the latest stable version of TF 2.6.0 ,less time difference is observed while running your code on colab ?Please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/cb415d8ae3f9d968ad60acb6e1f13352/conv2d_padding.ipynb)for reference & let us know if it helps?Thank you!

Sorry for the late reply. The times taken for running both the padding scheme are both shortened a lot in TF 2.6.0. But there is still a 3 times time difference observed on my machine. 
![image](https://user-images.githubusercontent.com/51120508/132947754-758f3143-01d3-441a-8d7e-ea08a9387262.png)

FYI, the CPU and GPU I used was Intel Xeon Processor (Skylake, IBRS) and Nvidia GeForce RTX 2080 Ti respectively.

**Comment by jvishnuvardhan** [comment]
@kelvinlyy I am seeing different computational time when compared to you. Can you please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/9557eb68974fe951f0f79881540fe6bc/conv2d_padding.ipynb) and let us know if I am missing anything here.

```
result = tf_conv2d_layer_valid(input_x)
Time taken: 0.005303621292114258

result = tf_conv2d_layer_same(input_x)
Time taken: 0.002894163131713867
```

**Comment by kelvinlyy** [comment]
> 
> 
> @kelvinlyy I am seeing different computational time when compared to you. Can you please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/9557eb68974fe951f0f79881540fe6bc/conv2d_padding.ipynb) and let us know if I am missing anything here.
> 
> ```
> result = tf_conv2d_layer_valid(input_x)
> Time taken: 0.005303621292114258
> 
> result = tf_conv2d_layer_same(input_x)
> Time taken: 0.002894163131713867
> ```

It is interesting. I ran the gist in your link but I got different results from yours.
![image](https://user-images.githubusercontent.com/51120508/133250897-d493adaa-cd13-444f-bc24-c1e19b10c8e7.png)

But when I ran the second time, it replicated your given runtime. 

I tried to restart the kernel and run again. It then took the approximate time I ran in the first time. I guess the problem can be reproduced if you first restart the kernel and then run the codes again. Can you please try that out?

**Comment by kelvinlyy** [comment]
> 
> 
> > @kelvinlyy I am seeing different computational time when compared to you. Can you please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/9557eb68974fe951f0f79881540fe6bc/conv2d_padding.ipynb) and let us know if I am missing anything here.
> > ```
> > result = tf_conv2d_layer_valid(input_x)
> > Time taken: 0.005303621292114258
> > 
> > result = tf_conv2d_layer_same(input_x)
> > Time taken: 0.002894163131713867
> > ```
> 
> It is interesting. I ran the gist in your link but I got different results from yours.
> ![image](https://user-images.githubusercontent.com/51120508/133250897-d493adaa-cd13-444f-bc24-c1e19b10c8e7.png)
> 
> But when I ran the second time, it replicated your given runtime.
> 
> I tried to restart the kernel and run again. It then took the approximate time I ran in the first time. I guess the problem can be reproduced if you first restart the kernel and then run the codes again. Can you please try that out?
> 
> FYI, I used "Factory reset runtime" to restart the kernel in Google Colab

I just spotted that if I run the conv2d with "same" and "valid" padding individually in 2 kernels, they took approximately the same time to compute. I guess the issue of performance difference should be fixed now. Thanks.

![image](https://user-images.githubusercontent.com/51120508/133253413-4e2898cc-8e5f-447d-9177-11664a790340.png)
